### PR TITLE
Improve Wan2.2 D2H on single and multi-host

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Summary
There were a few opportunities to improve D2H in Wan pipeline
- `torch.cat` memory allocations
- reading bf16 to host before conversion to uint8
- loading individual device shards to host buffers from python
On multi-host, it was even worse since we weren't taking advantage of 32 chips worth of PCIe bandwidth.

This PR makes a few improvements
- do uint8 conversion on device (requires https://github.com/tenstorrent/tt-metal/pull/42191 and https://github.com/tenstorrent/tt-metal/pull/42192 bugfixes)
- send device shards to cpu in single python call
- for multi-host, add binding to get specific shards from distributed host buffer. this is necessary because `get_device_tensors` on a distributed host buffer on multi-host causes a host all-gather, which is unnecessary and expensive in this use case

This saves 100ms e2e on 1xGLX and 500ms e2e on 4xGLX.

New perf results fast D2H

  | bf16 | uint8 |  
-- | -- | -- | --
1x | 169ms | 65ms |  
4x | 235ms | 111ms |  

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:cglagovich/wan_uint8_d2h)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:cglagovich/wan_uint8_d2h)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:cglagovich/wan_uint8_d2h)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:cglagovich/wan_uint8_d2h)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:cglagovich/wan_uint8_d2h)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:cglagovich/wan_uint8_d2h)

### CODE REVIEW

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
     bf16_tensor,
     fast_device_to_host,
     float32_tensor,
+    float_to_uint8,
```
Comment (sosborne-TT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py:40]: If bandwidth is still a bottleneck after these changes it might be worth looking at formats like YUV 4:2:0 instead of RGB which will give an additional 2x size reduction and may not affect accuracy (and the video encoder should support it).

Comment (cglagovichTT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:40]: Good point. 
I measured that the majority of time is now spent in AllGather and reassemble_2d.
I'll open an issue for us to continue investigating this.

Comment (sosborne-TT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:40]: Is there any reason not to use Gather instead of AllGather?

**Review by cfjchu — APPROVED** [review]

**Review by sosborne-TT — APPROVED** [review]

**Review by jonathansuTT — APPROVED** [review]

**Review by mbahnasTT — APPROVED** [review]

**Review by sosborne-TT — APPROVED** [review]

File: models/tt_dit/utils/tensor.py [hunk 2]
```diff
@@ -298,21 +298,159 @@ def _to_torch_zero_copy(t: ttnn.Tensor) -> torch.Tensor:
...
+    """On-device denormalization: map from [-1.0, 1.0] to [0.0, 1.0]."""
+    t = ttnn.to_layout(t, ttnn.TILE_LAYOUT)
+    t = ttnn.add(t, 1.0)
+    t = ttnn.multiply(t, 0.5)
+    t = ttnn.clamp(t, min=0.0, max=1.0)
+    t = ttnn.to_layout(t, ttnn.ROW_MAJOR_LAYOUT)
+    return t
+
+
+def float_to_uint8(t: ttnn.Tensor) -> ttnn.Tensor:
```
Comment (sadesoyeTT) [inline models/tt_dit/utils/tensor.py:331]: Can we make this a composition of float_to_unit_range to reduce duplicate code.

**Review by sadesoyeTT — APPROVED** [review]
